README.md
django-throttle
===============

.. image:: https://badge.fury.io/py/django-throttle.png
    :target: http://badge.fury.io/py/django-throttle
    
.. image:: https://travis-ci.org/marazmiki/django-throttle.png?branch=master
        :target: https://travis-ci.org/marazmiki/django-throttle

.. image:: https://coveralls.io/repos/marazmiki/django-throttle/badge.png?branch=master
  :target: https://coveralls.io/r/marazmiki/django-throttle?branch=master

.. image:: https://pypip.in/d/django-throttle/badge.png
        :target: https://pypi.python.org/pypi/django-throttle

Simple application for throttling views.


The license is MIT.




CHANGELOG.md
CHANGELOG
=========

0.2
---
* Added mixin for class based views
* Use markdown instead of RST markup

0.1
---
* Decorator

throttle/tests.py
from django import test
from django.conf.urls import url
from django.core.urlresolvers import reverse
from django.http import HttpResponse
from throttle.decorators import throttle
from throttle.views import ThrottleMixin
import time


class ThrottleTest(test.TestCase):
    """
    Throttle decorator test suite
    """
    urls = 'throttle.tests'
    name_map = {
        'default': 'test_default',
        'method': 'test_method',
        'duration': 'test_duration',
        'response': 'test_response',
        'response_callable': 'test_response_callable',
    }


    def request(self, url, method='post', **kwargs):
        """
        The helper function that emulates HTTP request to 
        Django views with given method
        """
        url = reverse(url)
        method = method.lower()
        return getattr(test.Client(), method)(url, **kwargs)

    def test_default(self):
        """
        Tests default usage
        """
        self.assertEquals(200, self.request(self.name_map['default']).status_code)
        self.assertEquals(403, self.request(self.name_map['default']).status_code)

    def test_method(self):
        """
        Tests that decorator applies to view for specified method and
        not applies for another
        """
        self.assertEquals(200, self.request(self.name_map['method'], method='GET').status_code)
        self.assertEquals(403, self.request(self.name_map['method'], method='GET').status_code)
        self.assertEquals(200, self.request(self.name_map['method'], method='POST').status_code)

    def test_response(self):
        """
        Tests custom response decorator argument
        """
        self.assertEquals(200, self.request(self.name_map['response']).status_code)
        self.assertEquals(401, self.request(self.name_map['response']).status_code)
        self.assertEquals(True, 'Response' in self.request(self.name_map['response']).content)

    def test_response_callable(self):
        """
        Tests custom response decorator argument
        """
        self.assertEquals(200, self.request(self.name_map['response_callable']).status_code)
        self.assertEquals(401, self.request(self.name_map['response_callable']).status_code)
        self.assertEquals(True, 'Request Response' in self.request(self.name_map['response_callable']).content)

    def test_duration(self):
        """
        Tests custom duration
        """
        self.assertEquals(200, self.request(self.name_map['duration']).status_code)
        self.assertEquals(200, self.request(self.name_map['duration']).status_code)


def index(request):
    """
    Test view function
    """
    return HttpResponse("Test view")


urlpatterns = [
    url(r'^$',  throttle(index), name='test_default'),
    url(r'^method/$',   throttle(index, method='GET'), name='test_method'),
    url(r'^duration/$', throttle(index, duration=0), name='test_duration'),
    url(r'^response/$', throttle(index, response=HttpResponse('Response', status=401)), name='test_response'),
    url(r'^response/callable/$', throttle(index, response=lambda request: HttpResponse('Request Response', status=401)), name='test_response_callable'),
]


try:
    from django.views.generic import View

except ImportError as e:
    print e
    pass

else:
    class IndexView(ThrottleMixin, View):
        """
        Class based view for mixin test
        """
        def get(self, request):
            return index(request)

        def post(self, request):
            return self.get(request)

    urlpatterns += [
        url(r'^cbv/$',  IndexView.as_view(), name='test_cbv_default'),
        url(r'^cbv/method/$',   IndexView.as_view(method='GET'), name='test_cbv_method'),
        url(r'^cbv/duration/$', IndexView.as_view(duration=0), name='test_cbv_duration'),
        url(r'^cbv/response/$', IndexView.as_view(response=HttpResponse('Response', status=401)), name='test_cbv_response'),
        url(r'^cbv/response/callable/$', IndexView.as_view(response=lambda request: HttpResponse('Request Response', status=401)), name='test_cbv_response_callable'),
    ]

    class ThrottleMixinTest(ThrottleTest):
        """
        ThrottleMixin test suite
        """
        name_map = {
        'default': 'test_cbv_default',
            'method': 'test_cbv_method',
            'duration': 'test_cbv_duration',
            'response': 'test_cbv_response',
            'response_callable': 'test_cbv_response_callable',
        }




throttle/decorators.py
from django.core.cache import cache
from django.http import HttpResponseForbidden, HttpResponse

def throttle(func, method='POST', duration=15, response=None):
    """
    This decorator is based on Django snippet #1573 code that 
    can be found at http://djangosnippets.org/snippets/1573/

    Simple usage

        @throttle
        def my_view(request):
            ""

    You can specify each of HTTP method

        @throttle(method='GET')
        def my_get_view(request)
            ""

    Custom

    """
    if response:
        if not isinstance(response, HttpResponse) and  not callable(response):
            raise TypeError, "The `response` keyword argument must " + \
                             "be a either HttpResponse instance or " + \
                             "callable with `request` argument.    "

    def inner(request, *args, **kwargs):
        if request.method == method:
            remote_addr = request.META.get('HTTP_X_FORWARDED_FOR') or \
                          request.META.get('REMOTE_ADDR')

            key = '{addr}.{path}'.format(addr=remote_addr, 
                                         path=request.get_full_path())
            if cache.get(key):
                if callable(response):
                    return response(request)

                elif response:
                    return response

                else:
                    return HttpResponseForbidden('Try slowing down a little.')

            cache.set(key, 1, duration)
        return func(request, *args, **kwargs)
    return inner


throttle/views.py
from django.core.cache import cache
from django.http import HttpResponseForbidden, HttpResponse


class ThrottleMixin(object):
    duration = 15
    method = 'POST'
    response = None

    def remote_addr(self):
        return self.request.META.get('HTTP_X_FORWARDED_FOR') or self.request.META.get('REMOTE_ADDR')

    def cache_key(self):
        return '{addr}.{path}'.format(addr=self.remote_addr(), 
            path=self.request.get_full_path())

    def dispatch(self, request, *args, **kwargs):
        self.request = request
        self.args = args
        self.kwargs = kwargs

        if (self.response and not isinstance(self.response, HttpResponse)
            and not callable(self.response)):
            raise TypeError("The `response` keyword argument must "
                            "be a either HttpResponse instance or "
                            "callable with `request` argument")

        if self.request.method == self.method:
            key = self.cache_key()

            if cache.get(key):
                if callable(self.response):
                    return self.response(request)

                elif self.response:
                    return self.response

                else:
                    return HttpResponseForbidden('Try slowing down a little.')

            cache.set(key, 1, self.duration)

        return super(ThrottleMixin, self).dispatch(request, *args, **kwargs)


def throttle(func, method='POST', duration=15, response=None):
    """
    This decorator is based on Django snippet #1573 code that 
    can be found at http://djangosnippets.org/snippets/1573/

    Simple usage

        @throttle
        def my_view(request):
            ""

    You can specify each of HTTP method

        @throttle(method='GET')
        def my_get_view(request)
            ""

    Custom

    """
    if response:
        if not isinstance(response, HttpResponse) and  not callable(response):
            raise TypeError, "The `response` keyword argument must " + \
                             "be a either HttpResponse instance or " + \
                             "callable with `request` argument.    "

    def inner(request, *args, **kwargs):
        if request.method == method:
            remote_addr = request.META.get('HTTP_X_FORWARDED_FOR') or \
                          request.META.get('REMOTE_ADDR')

            key = '{addr}.{path}'.format(addr=remote_addr, 
                                         path=request.get_full_path())
            if cache.get(key):
                if callable(response):
                    return response(request)

                elif response:
                    return response

                else:
                    return HttpResponseForbidden('Try slowing down a little.')

            cache.set(key, 1, duration)
        return func(request, *args, **kwargs)
    return inner
